### PR TITLE
feat: add verifier replay

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR creates a `verifier_replay` which validates whether verifier works correctly. But this requires you to fill data `kProof` and function body inside `GetInstanceColumnsVec()` manually.

This includes a bugfix of `SumOfProduct` and several refactorings.

### CODE REVIEW

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

File: tachyon/math/base/ring.h [hunk 1]
```diff
@@ -34,7 +34,6 @@ class Ring : public AdditiveGroup<R>, public MultiplicativeSemigroup<R> {
   constexpr static R SumOfProducts(const ContainerA& a, const ContainerB& b) {
     size_t size = std::size(a);
     CHECK_EQ(size, std::size(b));
-    CHECK_NE(size, size_t{0});
```
Comment (dongchangYoo) [inline tachyon/math/base/ring.h:37]: how about returning `R::Zero` if size == 0 ?

File: tachyon/c/zk/plonk/halo2/transcript_type.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    } else {
+      *reason = absl::Substitute("Unknown test set: $0", input);
+      return false;
+    }
+    return true;
+  }
+};
+
+}  // namespace base
+
```
Comment (dongchangYoo) [inline tachyon/c/zk/plonk/halo2/transcript_type.h:40]: Please remove a empty line between ends of namespaces.

**Review by dongchangYoo — CHANGES_REQUESTED** [review]

File: tachyon/math/base/ring.h (hunk 1 above)
Comment (chokobole) [reply tachyon/math/base/ring.h:37]: Okay!

**Review by ashjeong — COMMENTED** [review]
1. https://github.com/kroma-network/tachyon/pull/285/commits/ec369048aa7a56dcc88f5ef32974caa8fd27c7fe Does this now imply that all tests should be put in the tachyon namespace instead of the namespace for the respective files tested?

2. https://github.com/kroma-network/tachyon/pull/285/commits/e71b588d85a7c254cfb0ae7620ed9604af5bb6ef The subject implies the opposite of what you are trying to do -> change to "replace ToXPrimeField with c_cast and native_cast"

3. https://github.com/kroma-network/tachyon/pull/285/commits/0c24daf2dadf7978261cc79e9eef213a9a5b84ad 
change "add missing environmet check" to "add missing environment check"

3. https://github.com/kroma-network/tachyon/pull/285/commits/3a7decb2b2c84a15976fb646048ca2dd6316665c
change "allow arguments of SumOfProduct whose are empty containers" to "allow empty containers as arguments for SumOfProduct"

4. https://github.com/kroma-network/tachyon/pull/285/commits/04a59d2c671fcd06c1b357097ba8893f25249e6d
I think, in general, commits should have a more descriptive title explaining what the commit is about for easier understanding, so maybe changing this subject to "correct filename in comment" might be better?

**Review by Insun35 — APPROVED** [review]
LGTM

**Comment by chokobole** [comment]
> 5\. [04a59d2](https://github.com/kroma-network/tachyon/commit/04a59d2c671fcd06c1b357097ba8893f25249e6d)
>     I think, in general, commits should have a more descriptive title explaining what the commit is about for easier understanding, so maybe changing this subject to "correct filename in comment" might be better?

2. 3. 4. I'll apply

1. I intended to use the most common namespace! For example, after I change `namespace tachyon::c::math` to `namespace tachyon`, I can shorten `tachyon::math::bn254::G1PointXYZZ::Random()` to `math::bn254::G1PointXYZZ::Random()`.

5. "fix cpplint issue" is still good to me. Since cpp lint can correct many things and I don't want to separate commits because of this.

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Review by ashjeong — APPROVED** [review]
LGTM

**Pull request merged**